crates/core/component/stake/src/penalty.rs

```diff
@@ -12,52 +12,58 @@ use penumbra_asset::{
     balance::BalanceVar,
     Balance, Value, ValueVar, STAKING_TOKEN_ASSET_ID,
 };
-use penumbra_num::{fixpoint::bit_constrain, Amount, AmountVar};
+use penumbra_num::fixpoint::U128x128Var;
+use penumbra_num::{
+    fixpoint::{bit_constrain, U128x128},
+    Amount, AmountVar,
+};
 
 /// Tracks slashing penalties applied to a validator in some epoch.
 ///
-/// The penalty is represented as a fixed-point integer in bps^2 (denominator 10^8).
+/// Note(erwan): we are trying to get a sense of what a `U128x128` based penalty would
+/// look like. However, since this might be a bit too much for the undelegate claim
+/// circuit to handle, we can keep using an `Amount` in our back pocket.
+///
+/// Another thing to consider is to either limit the magnitude of the penalty, or change the
+/// API to propagate errors. Many operations are simply infallible if we assume that the penalty
+/// is at most 100%. It also aligns with how penalties are used in practice. TODO(erwan).
 #[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize, Default)]
 #[serde(try_from = "pbs::Penalty", into = "pbs::Penalty")]
-pub struct Penalty(pub u64);
+pub struct Penalty(pub U128x128);
 
 impl Penalty {
-    /// Create a `Penalty` from a percentage e.g.
-    /// `Penalty::from_percent(1)` is a 1% penalty.
-    /// `Penalty::from_percent(100)` is a 100% penalty.
     pub fn from_percent(percent: u64) -> Self {
-        Penalty::from_bps(percent * 100)
+        Penalty(U128x128::ratio(percent as u128, 100u128).expect("infaillible"))
     }
-
-    /// Create a `Penalty` from a basis point e.g.
-    /// `Penalty::from_bps(1)` is a 1 bps penalty.
-    /// `Penalty::from_bps(100)` is a 100 bps penalty.
     pub fn from_bps(bps: u64) -> Self {
-        Penalty(bps * 10_000)
+        Penalty(U128x128::ratio(bps as u128, 10000u128).expect("infaillible"))
     }
 
     /// Compound this `Penalty` with another `Penalty`.
     pub fn compound(&self, other: Penalty) -> Penalty {
-        // We want to compute q sth (1 - q) = (1-p1)(1-p2)
-        // q = 1 - (1-p1)(1-p2)
-        // but since each p_i implicitly carries a factor of 10^8, we need to divide by 10^8 after multiplying.
-        let one = 1_0000_0000u128;
-        let p1 = self.0 as u128;
-        let p2 = other.0 as u128;
-        let q = u64::try_from(one - (((one - p1) * (one - p2)) / 1_0000_0000))
-            .expect("value should fit in 64 bits");
+        /* Compounding:
+         * For two penalty rates p1 and p2, the compounded penalty rate q is:
+         *     `1 - q = (1-p1)(1-p2)``
+         *     `q = 1 - (1-p1)(1-p2)`
+         */
```

Comment:
do we want to do this, if we're using a `U128x128`? what if we just multiplied the penalties together?

 the previous representation did the (1-q) thing i think in order to make representing the fixed-point values easier, but if we are using a `U128x128` why not just make the penalty be multiplicative? wouldn't that have the least loss of precision?

Reply:
I think doing 1 - q vs q has no impact on precision at all in a fixed point representation

Reply:
The three subtractions become unnecessary, though, and it's simpler to reason about.

Reply:
Yeah, ultimately the 1 - is a matter of what the public interface for the component is, I agree that having rates be [0, 1) internally directly is simpler

Reply:
All the circuit needs to be doing is multiplying a number between 0 and 1 against the amount of tokens. We need to do that in the simplest way possible so we can reduce the risk of bugs.